lxd/resource_lxd_snapshot.go

```diff
@@ -0,0 +1,162 @@
+package lxd
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/lxc/lxd/shared/api"
+)
+
+func resourceLxdSnapshot() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceLxdSnapshotCreate,
+		Delete: resourceLxdSnapshotDelete,
+		Exists: resourceLxdSnapshotExists,
+		Read:   resourceLxdSnapshotRead,
+
+		Schema: map[string]*schema.Schema{
+			"container_name": &schema.Schema{
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+
+			"stateful": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+				ForceNew: true,
+			},
+
+			"creation_date": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceLxdSnapshotCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*LxdProvider).Client
+	remote := meta.(*LxdProvider).Remote
+
+	ctrName := d.Get("container_name").(string)
+	snapName := d.Get("name").(string)
+	stateful := d.Get("stateful").(bool)
+
+	// add a retry loop for creating snapshots
+	// sometimes a stateful snapshot creation will if attempting straight after container creation
+	var err error
+	var i int
+	for i = 0; i < 5; i++ {
+		var resp *api.Response
+		resp, err = client.Snapshot(ctrName, snapName, stateful)
+		if err != nil {
+			return err
+		}
+
+		// Wait for the LXC container to be created
+		err = client.WaitForSuccess(resp.Operation)
+		if err != nil {
+			if stateful && strings.Contains(err.Error(), "Dumping FAILED") {
+				log.Printf("[DEBUG] error creating stateful snapshot [%d]: %v", i, err)
+				time.Sleep(3 * time.Second)
+			} else {
+				return err
+			}
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		return fmt.Errorf("Failed to create snapshot after %d attempts, last error: %v", i, err)
+	}
+
+	snapId := NewSnapshotId(remote, ctrName, snapName)
+
+	d.SetId(snapId.String())
+
+	return resourceLxdSnapshotRead(d, meta)
+}
+
+func resourceLxdSnapshotRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*LxdProvider).Client
+	snapId := NewSnapshotIdFromResourceId(d.Id())
+
+	snap, err := client.SnapshotInfo(snapId.LxdId())
+	if err != nil {
+		if err.Error() == "not found" {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+	snapDump := spew.Sdump(snap)
+	log.Printf("[DEBUG] Snapshot Info: %s", snapDump)
```

Comment:
Would `%#v` work just as well here? Spew is great for developing, but might be a bit too verbose in the logs afterwards.